### DISCUSSION TITLE
How to set scrollbar to bottom for ScrollArea

### DISCUSSION THREAD

**Discussion by hasnainadil** [shadcn-ui/ui#4193] [Q&A]
I'm a bit new to react and nextjs development. I'm using nextjs14 for a dummy chatapp. I was trying to set the scrollbar to bottom when I enter the page so I tried this 

useEffect(() => {
        // Scroll to bottom of the scroll area
        if (scrollAreaRef.current) {
            console.log("scrolling to bottom");
            scrollAreaRef.current.scrollTop = scrollAreaRef.current.scrollHeight;
            console.log(scrollAreaRef.current.scrollHeight); //I'm getting 537 here
            console.log(scrollAreaRef.current.scrollTop); //I'm still getting 0 here height can't be set
        }
    }, [chatMessages, openedChat]);

<ScrollArea className="flex-1 p-4 overflow-y-auto" ref={scrollAreaRef}>
                        {chatMessages.map((message, index) => (
                            <div key={index} className="mb-4">
                                <p className="text-sm">{message}</p>
                            </div>
                        ))}
</ScrollArea >

this doesn't work at all but usual div works just fine 
regular div code:
<div className="flex-1 p-4 overflow-y-auto" ref={scrollAreaRef}>
                        {chatMessages.map((message, index) => (
                            <div key={index} className="mb-4">
                                <p className="text-sm">{message}</p>
                            </div>
                        ))}
</div>
this code above works fine with the code in useEffect
I also tired to use id to get the scroll area but not working. How can I set the scrolltop to bottom of the ScrollArea with shadcn ScrollArea component. 
Thanks in advance.

**Comment by abhayclasher** [comment]
hey, for `ScrollArea` manual `scrollTop` is a pain. easiest way is to just put a dummy `div` at the very bottom of your message list and use a ref on it.

```tsx
const bottomRef = useRef<HTMLDivElement>(null)

useEffect(() => {
  bottomRef.current?.scrollIntoView({ behavior: "smooth" })
}, [chatMessages])

// ... inside your ScrollArea
{chatMessages.map(...)}
<div ref={bottomRef} />
```

this way it always jumps to the end whenever `chatMessages` updates. way cleaner 👍

**Comment by Manavarya09** [comment]
The issue is that `ScrollArea` wraps your content in an internal viewport, 
so setting `scrollTop` on the outer ref won't work.

Use a bottom anchor ref instead:

\`\`\`tsx
const bottomRef = useRef<HTMLDivElement>(null);

useEffect(() => {
  bottomRef.current?.scrollIntoView({ behavior: "smooth" });
}, [messages]);

return (
  <ScrollArea className="h-[400px]">
    {messages.map((msg) => (
      <div key={msg.id}>{msg.content}</div>
    ))}
    <div ref={bottomRef} />
  </ScrollArea>
);
\`\`\`

This works because `scrollIntoView()` navigates the nearest scrollable 
ancestor regardless of component wrapping. Much cleaner than manual 
scroll position calculation.

**Comment by hesamhadadi** [comment]
The reason `scrollIntoView` and `scrollTop` don't work with shadcn's `ScrollArea` is that Radix wraps your content in a custom viewport — the actual scrollable element is **inside** `ScrollArea`, not the component itself. You need to grab the viewport ref and scroll it directly.

## Solution

```tsx
'use client';
import { useRef, useEffect } from 'react';
import { ScrollArea } from '@/components/ui/scroll-area';

export default function Chat({ messages }: { messages: string[] }) {
  const viewportRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const viewport = viewportRef.current;
    if (!viewport) return;
    viewport.scrollTo({
      top: viewport.scrollHeight,
      behavior: 'smooth', // omit for instant scroll
    });
  }, [messages]); // re-runs every time a new message arrives

  return (
    <ScrollArea className="h-[500px]" viewportRef={viewportRef}>
      {messages.map((m, i) => (
        <div key={i} className="p-3">{m}</div>
      ))}
    </ScrollArea>
  );
}
```

## If your shadcn version doesn't expose `viewportRef`

Older shadcn copies don't forward it. Patch `components/ui/scroll-area.tsx`:

```tsx
const ScrollArea = React.forwardRef<
  React.ElementRef<typeof ScrollAreaPrimitive.Root>,
  React.ComponentPropsWithoutRef<typeof ScrollAreaPrimitive.Root> & {
    viewportRef?: React.Ref<HTMLDivElement>;
  }
>(({ className, children, viewportRef, ...props }, ref) => (
  <ScrollAreaPrimitive.Root
    ref={ref}
    className={cn('relative overflow-hidden', className)}
    {...props}
  >
    <ScrollAreaPrimitive.Viewport
      ref={viewportRef}
      className="h-full w-full rounded-[inherit]"
    >
      {children}
    </ScrollAreaPrimitive.Viewport>
    <ScrollBar />
    <ScrollAreaPrimitive.Corner />
  </ScrollAreaPrimitive.Root>
));
```

## Why it works

`ScrollArea` is a Radix primitive that creates a custom div with `overflow: scroll` *inside itself*. Calling `scrollTop` on the outer `ScrollArea` does nothing because that element has `overflow: hidden`. The viewport is the only element that actually scrolls.

For your specific use case (chat scrolling to bottom on mount + on new message), this is the standard pattern and works reliably.
